gh-99726: Adds os.statx function and associated constants (#99755)
python/cpython: Doc/library/os.rst

      ``follow_symlinks=False`` had been specified instead of raising an error.

.. function:: statx(path, mask, *, dir_fd=None, follow_symlinks=True, flags=0)

   Get selected fields of the status of a file or a file descriptor. Perform the
   equivalent of a :c:func:`statx` system call on the given path. *path* may be
   specified as either a string or bytes -- directly or indirectly through the
   :class:`PathLike` interface -- or as an open file descriptor. Return a
   :class:`stat_result` object.

   *mask* is a bitwise combination of the ``STATX_*`` attributes in the
   :mod:`stat` module, indicating which fields the caller intends to use. Note
   that the set of fields returned may differ from what's requested, if the
   operating system or file system does not support the metadata, or if it can
   provide additional fields with no extra effort.

>    This function normally follows symlinks; to stat a symlink add the argument
>    ``follow_symlinks=False``.
> 

Author: ```suggestion
   This function normally follows symlinks; to stat a symlink add the argument
   ``follow_symlinks=False``.

   This function is present even if the operating system does not support
   ``statx``. In this case, the *mask* argument is ignored and a regular
   :func:`stat`, :func:`lstat` or :func:`fstat` call will be used.
   The :attr:`stat_result.stx_mask` field is always present, and is initialized
   with a meaningful value for regular stat calls.

```

   This function can support :ref:`specifying a file descriptor <path_fd>` and
   :ref:`not following symlinks <follow_symlinks>`.

   .. index:: module: stat

   Example::

      >>> import os, stat
      >>> statinfo = os.statx('somefile.txt', stat.STATX_SIZE)
      >>> statinfo
      os.stat_result(st_mode=33188, st_ino=0, st_dev=0,
      st_nlink=1, st_uid=0, st_gid=0, st_size=264, st_atime=0,
      st_mtime=0, st_ctime=0)
      >>> statinfo.stx_mask & stat.STATX_SIZE
      512
      >>> statinfo.st_size
      264

   .. seealso::

      :func:`stat`, :func:`fstat` and :func:`lstat` functions.

   .. versionadded:: 3.12
      Added ``statx`` function.

.. class:: stat_result

   Object whose attributes correspond roughly to the members of the
   :c:type:`stat` structure. It is used for the result of :func:`os.stat`,
   :func:`os.fstat`, :func:`os.lstat` and :func:`os.statx`.

>    Attributes that are optional under :func:`os.stax` calls are marked

Author: ```suggestion
   Attributes that are optional under :func:`os.stax` calls are documented
```

   with the mask value that will be present in :attr:`stx_mask` when the
   value is present. This can be tested even for regular ``stat`` calls,
   as ``stx_mask`` will be initialized with a suitable value even if the
   underlying system call uses regular ``stat``.

   Attributes may be missing from this struct if they are not supported by
   the underlying operating system. When :func:`os.statx` is supported, all
   ``stx_*`` attributes are present, but their ``st_*`` equivalents may not
   be. :attr:`stx_mask` is always present, even if ``statx`` is not supported.

   Attributes:

   .. attribute:: st_mode

      File mode: file type and file mode bits (permissions).

>       This field is set when ``stx_mask`` contains

Author: ```suggestion
      This field is valid when ``stx_mask`` contains
```

   :mod:`stat` module, indicating which fields the caller intends to use. Note
   that the set of fields returned may differ from what's requested, if the
   operating system or file system does not support the metadata, or if it can
   provide additional fields with no extra effort.

   This function normally follows symlinks; to stat a symlink add the argument
   ``follow_symlinks=False``.

   This function can support :ref:`specifying a file descriptor <path_fd>` and
   :ref:`not following symlinks <follow_symlinks>`.

   On Windows, passing ``follow_symlinks=False`` will disable following all
   name-surrogate reparse points, which includes symlinks and directory
   junctions. Other types of reparse points that do not resemble links or that
   the operating system is unable to follow will be opened directly. When
   following a chain of multiple links, this may result in the original link
   being returned instead of the non-link that prevented full traversal. To
   obtain stat results for the final path in this case, use the
   :func:`os.path.realpath` function to resolve the path name as far as
   possible and call :func:`lstat` on the result. This does not apply to

>    dangling symlinks or junction points, which will raise the usual exceptions.

Reviewer: I'd prefer for the `follow_symlinks` behavior to be discussed in the ["Files and Directories"](https://docs.python.org/3/library/os.html#files-and-directories) section instead of in the `stat()` and `statx()` documentation. If we add `follow_symlinks` support to other os functions (e.g. `os.access()`), the behavior had better be consistent with `os.stat()`. For example:

```text
* **not following symlinks:** If *follow_symlinks* is
  ``False``, and the last element of the path to operate on is a symbolic link,
  the function will operate on the symbolic link itself rather than the file
  pointed to by the link.  (For POSIX systems, Python will call the ``l...``
  variant of the function.)

  You can check whether or not *follow_symlinks* is supported for a particular
  function on your platform using :data:`os.supports_follow_symlinks`.
  If it's unavailable, using it will raise a :exc:`NotImplementedError`.

  On Windows, ``follow_symlinks`` applies to name-surrogate reparse points,
  including symlinks and junctions. This type of reparse point is a symbolic
  link to another path on the system.  All other types of reparse point (e.g.
  tiered storage) are always traversed, except that reparse points that are
  not supported by the system are operated on directly.

  If a path is either a symlink or a junction that cannot be traversed to the
  final path, :func:`os.path.realpath` can be used to resolve as much of the
  target path as possible.
```
This example simplifies the discussion a bit. The chain of links discussion is moot. `stat()` will fail if the path is a symlink or junction that can't be resolved, regardless of whether it chains to an unsupported reparse point or if some target device/file/directory doesn't exist or is inaccessible. Other reparse points are black boxes. If they fail for any reason, there's nothing we can do to resolve the path beyond the reparse point.

        <https://msdn.microsoft.com/en-us/library/aa363788>`_ on
        Windows

>       This field is set when ``stx_mask`` contains :data:`stat.STATX_INO`.

Author: ```suggestion
      This field is valid when ``stx_mask`` contains :data:`stat.STATX_INO`.
```

   .. attribute:: st_dev

      Identifier of the device on which this file resides.

>       On Windows, this field is set with :data:`stat.STATX_INO`.

Author: ```suggestion
      On Windows, this field is valid with :data:`stat.STATX_INO`.
```

      Other platforms always set this value.

   .. attribute:: st_nlink

      Number of hard links.

>       This field is set when ``stx_mask`` contains :data:`stat.STATX_NLINK`.

Author: ```suggestion
      This field is valid when ``stx_mask`` contains :data:`stat.STATX_NLINK`.
```

   .. attribute:: st_uid

      User identifier of the file owner.

>       This field is set when ``stx_mask`` contains :data:`stat.STATX_UID`.

Author: ```suggestion
      This field is valid when ``stx_mask`` contains :data:`stat.STATX_UID`.
```

   .. attribute:: st_gid

      Group identifier of the file owner.

>       This field is set when ``stx_mask`` contains :data:`stat.STATX_GID`.

Author: ```suggestion
      This field is valid when ``stx_mask`` contains :data:`stat.STATX_GID`.
```

   .. attribute:: st_size

      Size of the file in bytes, if it is a regular file or a symbolic link.
      The size of a symbolic link is the length of the pathname it contains,
      without a terminating null byte.

>       This field is set when ``stx_mask`` contains :data:`stat.STATX_SIZE`.

Author: ```suggestion
      This field is valid when ``stx_mask`` contains :data:`stat.STATX_SIZE`.
```

   .. attribute:: stx_mask

>       Flags indicating which values were set. :func:`os.statx` allows specifying
>       a mask, though the result may include more or less than requested. Other
>       ``stat`` functions set a default value representing the information they
>       return.

Author: ```suggestion
      Flags indicating which values in the result are valid. :func:`os.statx`
      allows specifying a mask, though the result may include more or less
      than requested. Other ``stat`` functions set a default value representing
      the information they return.
```

   Timestamps:

   .. attribute:: st_atime

      Time of most recent access expressed in seconds.

>       This field is set when ``stx_mask`` contains :data:`stat.STATX_ATIME`.

Author: ```suggestion
      This field is valid when ``stx_mask`` contains :data:`stat.STATX_ATIME`.
```

   .. attribute:: st_mtime

      Time of most recent content modification expressed in seconds.

>       This field is set when ``stx_mask`` contains :data:`stat.STATX_MTIME`.

Author: ```suggestion
      This field is valid when ``stx_mask`` contains :data:`stat.STATX_MTIME`.
```

   .. attribute:: st_ctime

      Platform dependent:

      * the time of most recent metadata change on Unix,
      * the time of creation on Windows, expressed in seconds, except

>         when :data:`stat.STATX_CTIME` is in :attr:`stx_mask`, in which
>         case this is the time of the most recent metadata change

Author: ```suggestion
        when :data:`stat.STATX_CTIME` is in :attr:`stx_mask`, in which
        case this is the time of the most recent metadata change.
        Regular stat calls will set this field without setting the mask.
```

>       This field is set when ``stx_mask`` contains :data:`stat.STATX_CTIME`.

Author: ```suggestion
      This field is valid when ``stx_mask`` contains :data:`stat.STATX_CTIME`.
```

   .. attribute:: st_atime_ns

      Time of most recent access expressed in nanoseconds as an integer.

>       This field is set when ``stx_mask`` contains :data:`stat.STATX_ATIME`.

Author: ```suggestion
      This field is valid when ``stx_mask`` contains :data:`stat.STATX_ATIME`.
```

   .. attribute:: st_mtime_ns

      Time of most recent content modification expressed in nanoseconds as an
      integer.

>       This field is set when ``stx_mask`` contains :data:`stat.STATX_MTIME`.

Author: ```suggestion
      This field is valid when ``stx_mask`` contains :data:`stat.STATX_MTIME`.
```

   .. attribute:: st_ctime_ns

      Platform dependent:

      * the time of most recent metadata change on Unix,
      * the time of creation on Windows, expressed in nanoseconds as an
        integer, except when :data:`stat.STATX_CTIME` is in :attr:`stx_mask`,

>         in which case this is the time of the most recent metadata change

Author: ```suggestion
        in which case this is the time of the most recent metadata change.
        Regular stat calls will set this field without setting the mask.
```

>       This field is set when ``stx_mask`` contains :data:`stat.STATX_CTIME`.

Author: ```suggestion
      This field is valid when ``stx_mask`` contains :data:`stat.STATX_CTIME`.
```

   .. attribute:: st_birthtime

      Time of file creation, if available. The attribute may not be present if
      your operating system does not support the field.

>       This field is set with :data:`stat.STATX_BTIME`, but is only present when

Author: ```suggestion
      This field is valid with :data:`stat.STATX_BTIME`, but is only present when
```

      supported by regular ``stat`` calls. See also: :attr:`stx_btime`.

   .. attribute:: stx_btime

      Alias of :attr:`st_birthtime` that is always present when :func:`statx`
      is supported. If ``st_birthtime`` is also present, its value will be
      identical.

>       This field is set when ``stx_mask`` contains :data:`stat.STATX_BTIME`.

Author: ```suggestion
      This field is valid when ``stx_mask`` contains :data:`stat.STATX_BTIME`.
```

   .. attribute:: stx_btime_ns

      Time of file creation expressed in nanoseconds as an integer.

>       This field is set when ``stx_mask`` contains :data:`stat.STATX_BTIME`.

Author: ```suggestion
      This field is valid when ``stx_mask`` contains :data:`stat.STATX_BTIME`.
```

        <https://msdn.microsoft.com/en-us/library/aa363788>`_ on

>         Windows

Reviewer 2: This looks incomplete in the source here, i assume this renders to something different?

Author: There's one more line that's not visible in the diff, because I didn't touch these lines. You can see it at https://docs.python.org/3/library/os.html#os.stat_result.st_ino where I think it looks fine

      Number of 512-byte blocks allocated for file.
      This may be smaller than :attr:`st_size`/512 when the file has holes.

>       This field is set with :data:`stat.STATX_BLOCKS`, but is only present

Author: ```suggestion
      This field is valid with :data:`stat.STATX_BLOCKS`, but is only present
```

      when supported by regular ``stat`` calls. See also: :attr:`stx_blocks`.

   .. attribute:: st_blksize

      "Preferred" blocksize for efficient file system I/O. Writing to a file in
      smaller chunks may cause an inefficient read-modify-rewrite.

>       This field is set with :data:`stat.STATX_BLOCKSIZE`, but is only present

Author: ```suggestion
      This field is valid with :data:`stat.STATX_BLOCKSIZE`, but is only present
```

      when supported by regular ``stat`` calls. See also: :attr:`stx_blksize`.

   .. attribute:: st_rdev

      Type of device if an inode device.

>       This field is always set when appropriate, but is only present when

Author: ```suggestion
      This field is always valid when present, but is only present when
```

      :c:func:`GetFileInformationByHandle`. See the ``FILE_ATTRIBUTE_*``
      constants in the :mod:`stat` module.

>       This field is requested with :data:`stat.STATX_TYPE`.

Author: ```suggestion
      This field is valid when ``stx_mask`` contains :data:`stat.STATX_TYPE`.
```

   .. attribute:: st_reparse_tag

      When :attr:`st_file_attributes` has the ``FILE_ATTRIBUTE_REPARSE_POINT``
      set, this field contains the tag identifying the type of reparse point.
      See the ``IO_REPARSE_TAG_*`` constants in the :mod:`stat` module.

>       This field is requested with :data:`stat.STATX_TYPE`.

Author: ```suggestion
      This field is valid when ``stx_mask`` contains :data:`stat.STATX_TYPE`.
```

      files as :const:`S_IFCHR`, :const:`S_IFIFO` or :const:`S_IFBLK`
      as appropriate.

   .. versionchanged:: 3.12

>       Added the :attr:`stx_mask` member along with :func:`statx`.

Author: ```suggestion
      Added :attr:`stx_mask` and other ``stx_*`` members along with :func:`statx`.
```
